### ISSUE TITLE
Acquiring "is_grad_enabled()" inside an autograd function 

### ISSUE THREAD

**Issue by zbh2047** [pytorch/pytorch#56370]
## 🚀 Feature
Acquiring whether gradient is enabled in a custom autograd function would be very useful. Currently such information can not be obtained.

## Motivation
Custom autograd functions (i.e. by inheriting `torch.autograd.Function`) is frequently used in Pytorch. To write a custom autograd function, we will implement two static methods: `forward` and `backward`. Currently in the `forward` function, the gradient is automatically disabled and `is_grad_enabled()` always returns false. So we can not know whether `backward` will be called later.

In some situations such information would be very useful, especially for speeding up the calculation. For example, consider the `max` function in Pytorch. If we need to calulate the gradient of `max`, we must save the index that achieves maximum, since the gradient is non-zero only at the saved index. On the other hand, if we know in advance that we do not need to calculate gradient, then the implementation can be significantly simplified so that only a max-reduction is needed. This can be much faster than calculating the maximum and the index that achieves maximum.

Another situation is the `sort` function that sorts an array. If we know in advance that we do not need to calculate the gradient, then we do not need to save the rank of each element. This again speeds up the running time significantly.

## Pitch

It would be very nice if Pytorch can provide a function to acquire whether gradient is enabled inside the `forward` method of a custom autograd function.

## Alternatives

Currently I has no idea how to achieve above goal. So the forward implementation is slow even if I know I do not need to calcualte gradient.

## Additional context

Here is the pseudo-code about the feature metioned above (using the `max` example):

```
class Max(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        if not torch.is_grad_enabled(): # Currently it always returns false.
            return x.max() # only calculate the maximum, which is fast
        y, index = x.max(dim=0) # calculate both the maximum and the index, which is slow
        ctx.save_for_backward(x, index)
        return y

    @staticmethod
    def backward(ctx, grad):
        x, index = ctx.saved_tensors
        res = torch.zeros_like(x)
        res[index] = grad
        return res

```

**Comment by ezyang** [comment]
This seems like a very reasonable request.

**Comment by ak-source** [comment]
solution by is_grad_enabled is not ideal, as it won't catch uses with detached tensors / frozen parameters

I'd suggest supporting optional "fast path" instead, like:
```
@staticmethod
def forward_nograd(x):
  return x.max()
```
... and detect uses on tensors having requires_grad==False in addition to global no_grad state

**Comment by albanD** [comment]
At that point in the code, we have all the information we need to know if the graph is going to get built or not (because both of the no_grad state and the input Tensor's requires_grad fields).
So we could have a `ctx.needs_grad` available in the forward (similar to the `ctx.needs_input_grad` available in the backward). This will tell you if the graph is being built or not.
Would that solve your problem?

**Comment by zbh2047** [comment]
Yes, any approach above can solve this problem.

**Comment by zbh2047** [comment]
If we can check the need_grad property to speed up calculation, I think pytorch might also consider providing an interface without returning index for some important built-in functions. These functions may include: min, max, cummax, sort, topk. Other functions related to max operator can also be implemented based on this feature (but without changing interface), such as Maxpool, vector L_infinity norm, maxtrix 1-norm or infinity-norm. The fast implementation can also reduce half of the memory cost since we do not need to save the index.
